File: src/test/java/com/sashashpota/twitter/PostTweetActionTest.java
package com.sashashpota.twitter;

import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.Presentation;
import com.intellij.openapi.editor.Caret;
import com.intellij.openapi.editor.CaretModel;
import com.intellij.openapi.editor.Editor;
import org.junit.Test;

import static com.intellij.openapi.actionSystem.CommonDataKeys.EDITOR;
import static org.hamcrest.core.Is.is;
import static org.junit.Assert.assertThat;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.mock;

public class PostTweetActionTest {
    @Test
    public void shouldUpdateVisibility() {
        PostTweetAction action = new PostTweetAction();
        AnActionEvent event = mock(AnActionEvent.class);
        Editor editor = mock(Editor.class);
        given(event.getData(EDITOR)).willReturn(editor);
        CaretModel caretModel = mock(CaretModel.class);
        given(editor.getCaretModel()).willReturn(caretModel);
        Caret currentCaret = mock(Caret.class);
        given(caretModel.getCurrentCaret()).willReturn(currentCaret);
        given(currentCaret.getSelectedText()).willReturn("selected text");
        Presentation presentation = new Presentation();
        given(event.getPresentation()).willReturn(presentation);

        action.update(event);

        assertThat(presentation.isVisible(), is(true));
    }

    @Test
    public void shouldUpdateVisibilityGivenBlankSelectedText() {
        PostTweetAction action = new PostTweetAction();
        AnActionEvent event = mock(AnActionEvent.class);
        Editor editor = mock(Editor.class);
        given(event.getData(EDITOR)).willReturn(editor);
        CaretModel caretModel = mock(CaretModel.class);
        given(editor.getCaretModel()).willReturn(caretModel);
        Caret currentCaret = mock(Caret.class);
        given(caretModel.getCurrentCaret()).willReturn(currentCaret);
        given(currentCaret.getSelectedText()).willReturn("     ");
        Presentation presentation = new Presentation();
        given(event.getPresentation()).willReturn(presentation);

        action.update(event);

        assertThat(presentation.isVisible(), is(false));
    }

    @Test
    public void shouldUpdateVisibilityGivenNullEditor() {
        PostTweetAction action = new PostTweetAction();
        AnActionEvent event = mock(AnActionEvent.class);
        Presentation presentation = new Presentation();
        given(event.getPresentation()).willReturn(presentation);

        action.update(event);

        assertThat(presentation.isVisible(), is(false));
    }
}

File: src/test/kotlin/com/sashashpota/twitter/PostTweetActionTest.kt
package com.sashashpota.twitter

import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys.EDITOR
import com.intellij.openapi.actionSystem.Presentation
import com.intellij.openapi.editor.Caret
import com.intellij.openapi.editor.CaretModel
import com.intellij.openapi.editor.Editor
import org.junit.Assert.assertEquals
import org.junit.Test
import org.mockito.BDDMockito.given
import org.mockito.Mockito.mock

class PostTweetActionTest {
    @Test
    fun shouldUpdateVisibility() {
        val action = PostTweetAction()
        val event = mock(AnActionEvent::class.java)
        val editor = mock(Editor::class.java)
        given(event.getData(EDITOR)).willReturn(editor)
        val caretModel = mock(CaretModel::class.java)
        given(editor.caretModel).willReturn(caretModel)
        val currentCaret = mock(Caret::class.java)
        given(caretModel.currentCaret).willReturn(currentCaret)
        given(currentCaret.selectedText).willReturn("selected text")
        val presentation = Presentation()
        given(event.presentation).willReturn(presentation)

        action.update(event)

        assertEquals(presentation.isVisible, true)
    }

    @Test
    fun shouldUpdateVisibilityGivenBlankSelectedText() {
        val action = PostTweetAction()
        val event = mock(AnActionEvent::class.java)
        val editor = mock(Editor::class.java)
        given(event.getData(EDITOR)).willReturn(editor)
        val caretModel = mock(CaretModel::class.java)
        given(editor.caretModel).willReturn(caretModel)
        val currentCaret = mock(Caret::class.java)
        given(caretModel.currentCaret).willReturn(currentCaret)
        given(currentCaret.selectedText).willReturn(" ")
        val presentation = Presentation()
        given(event.presentation).willReturn(presentation)

        action.update(event)

        assertEquals(presentation.isVisible, false)
    }

    @Test
    fun shouldUpdateVisibilityGivenNullEditor() {
        val action = PostTweetAction()
        val event = mock(AnActionEvent::class.java)
        val presentation = Presentation()
        given(event.presentation).willReturn(presentation)

        action.update(event)

        assertEquals(presentation.isVisible, false)
    }
}